File: ai-solutions/01/gpt-5.1-codex/python/main.py
#!/usr/bin/env python3
"""
Advent of Code 2025 - Day 01
GPT-5.1 Codex solution (Python)
"""

import sys
from pathlib import Path
from typing import Iterable, List, Tuple

MODULO = 100


def parse_moves(lines: Iterable[str]) -> List[Tuple[str, int]]:
    """Convert each rotation string like 'R17' into (direction, distance)."""
    moves: List[Tuple[str, int]] = []
    for raw in lines:
        line = raw.strip()
        if not line:
            continue
        direction = line[0]
        distance = int(line[1:])
        moves.append((direction, distance))
    return moves


def count_zero_hits(position: int, distance: int, direction: str) -> int:
    """
    Count how many times a rotation causes the pointer to click onto 0.
    Works by finding the first step that lands on 0, then counting every 100 clicks.
    """
    if distance <= 0:
        return 0

    if direction == "R":
        first = (MODULO - position) % MODULO
    else:
        first = position % MODULO

    if first == 0:
        first = MODULO

    if distance < first:
        return 0

    return 1 + (distance - first) // MODULO


def solve(moves: List[Tuple[str, int]]) -> Tuple[int, int]:
    """Return (part1, part2) passwords."""
    position = 50
    part1 = 0
    part2 = 0

    for direction, distance in moves:
        part2 += count_zero_hits(position, distance, direction)
        if direction == "R":
            position = (position + distance) % MODULO
        else:
            position = (position - distance) % MODULO

        if position == 0:
            part1 += 1

    return part1, part2


def main() -> None:
    if len(sys.argv) < 2:
        print("Usage: python main.py <input_file>", file=sys.stderr)
        sys.exit(1)

    input_path = Path(sys.argv[1])
    if not input_path.exists():
        print(f"Input file not found: {input_path}", file=sys.stderr)
        sys.exit(1)

    moves = parse_moves(input_path.read_text().splitlines())
    part1, part2 = solve(moves)

    print(part1)
    print(part2)


if __name__ == "__main__":
    main()

File: ai-solutions/01/gpt-5.1-codex/rust/src/main.rs
use std::env;
use std::fs;
use std::path::Path;

const MODULO: i64 = 100;

fn parse_moves(input: &str) -> Vec<(char, i64)> {
    input
        .lines()
        .filter_map(|line| {
            let trimmed = line.trim();
            if trimmed.is_empty() {
                return None;
            }
            let mut chars = trimmed.chars();
            let dir = chars.next().unwrap_or('R');
            let distance = chars.as_str().parse::<i64>().ok()?;
            Some((dir, distance))
        })
        .collect()
}

fn count_zero_hits(position: i64, distance: i64, direction: char) -> i64 {
    if distance <= 0 {
        return 0;
    }

    let mut first = if direction == 'R' {
        (MODULO - position).rem_euclid(MODULO)
    } else {
        position.rem_euclid(MODULO)
    };

    if first == 0 {
        first = MODULO;
    }

    if distance < first {
        0
    } else {
        1 + (distance - first) / MODULO
    }
}

fn solve(moves: &[(char, i64)]) -> (i64, i64) {
    let mut position = 50_i64;
    let mut part1 = 0_i64;
    let mut part2 = 0_i64;

    for &(direction, distance) in moves {
        part2 += count_zero_hits(position, distance, direction);

        if direction == 'R' {
            position = (position + distance).rem_euclid(MODULO);
        } else {
            position = (position - distance).rem_euclid(MODULO);
        }

        if position == 0 {
            part1 += 1;
        }
    }

    (part1, part2)
}

fn run() -> Result<(), String> {
    let mut args = env::args().skip(1);
    let input_path = args
        .next()
        .ok_or_else(|| "Usage: cargo run -- <input_file>".to_string())?;

    let absolute = Path::new(&input_path);
    if !absolute.exists() {
        return Err(format!("Input file not found: {}", absolute.display()));
    }

    let content = fs::read_to_string(absolute)
        .map_err(|err| format!("Failed to read {}: {}", absolute.display(), err))?;

    let moves = parse_moves(&content);
    let (part1, part2) = solve(&moves);
    println!("{}", part1);
    println!("{}", part2);
    Ok(())
}

fn main() {
    if let Err(err) = run() {
        eprintln!("{}", err);
        std::process::exit(1);
    }
}






